DiscountDiscover/View Controllers/DealsViewController.m

```diff
@@ -40,34 +41,42 @@ - (void)viewDidLayoutSubviews {
     layout.minimumInteritemSpacing = 10;
     layout.minimumLineSpacing = 10;
     const int sectionInsets = 20;
+    const int dealsPerLine = 2;
     
-    const CGFloat itemWidth = (self.collectionView.frame.size.width - layout.minimumInteritemSpacing - sectionInsets) / 2;
+    const CGFloat itemWidth = (self.collectionView.frame.size.width - layout.minimumInteritemSpacing - sectionInsets) / dealsPerLine;
     const CGFloat itemHeight = itemWidth * 1.5;
     layout.itemSize = CGSizeMake(itemWidth, itemHeight);
 }
 
 - (void)fetchDeals {
-//    [self.activityIndicator startAnimating];
+    APIManager *manager = [[APIManager alloc] init];
     __weak DealsViewController *weakSelf = self;
-    APIManager *manager = [APIManager new];
     [manager fetchDeals:^(NSArray<Deal *> *deals, NSError *error) {
+        DealsViewController *strongSelf = weakSelf;
         if (error != nil) {
-            UIAlertController *networkAlert = [UIAlertController alertControllerWithTitle:@"Cannot Get Deals" message:@"The Internet connection appears to be offline." preferredStyle:(UIAlertControllerStyleAlert)];
-            UIAlertAction *reloadAction = [UIAlertAction actionWithTitle:@"Try Again" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
-                [weakSelf fetchDeals];
-            }];
-            [networkAlert addAction:reloadAction];
-            [self presentViewController:networkAlert animated:YES completion:^{}];
+            [strongSelf showNetworkErrorAlert];
         }
         else {
-            self.deals = deals;
-            [self.collectionView reloadData];
+            strongSelf.deals = deals;
+            [strongSelf.collectionView reloadData];
         }
-//        [self.activityIndicator stopAnimating];
-        [self.refreshControl endRefreshing];
+        [strongSelf.refreshControl endRefreshing];
     }];
 }
 
+- (void)showNetworkErrorAlert {
+    UIAlertController *networkAlert = [UIAlertController alertControllerWithTitle:@"Cannot Get Deals" message:@"The Internet connection appears to be offline." preferredStyle:(UIAlertControllerStyleAlert)];
+    
+    __weak DealsViewController *weakSelf = self;
+    UIAlertAction *reloadAction = [UIAlertAction actionWithTitle:@"Try Again" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
+        DealsViewController *strongSelf = weakSelf;
```

Comment:
To get a strong reference from the weak reference you would usually do something like this:
```
__weak typeof(self) weakSelf = self;
UIAlertAction *reloadAction = [UIAlertAction actionWithTitle:@"Try Again" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
   __strong typeof(self) strongSelf = self;
  if (strongSelf) {
    // do stuff
  }
}];
```
Notice the `__strong` directive, and the nil check because the weak reference might be nil by the time the block is executed. Using `typeof(self)` instead of the actual class makes this more generic (and reusable as a code snippet for example).

Reply:
If you don't care if `weakSelf` is nil when the block is executed (ie you're just calling a method like `fetchDeals` here) you can skip the strong self part and use weakSelf directly in the block like `[weakSelf fetchDeals]`, in Objective-C you can safely invoke a method in a nil reference and nothing happens. The `strongSelf` reference is necessary if you need to get some data from self inside the block, like accessing some of the properties.